api/buckets/service.go

```diff
@@ -1702,78 +1715,82 @@ func (s *Service) unpinPath(ctx context.Context, path path.Path) error {
 // pinBlocks pin the provided blocks to the IPFS node, and accounts to the
 // account/user buckets total size quota.
 func (s *Service) pinBlocks(ctx context.Context, nodes []ipld.Node) error {
-	if err := s.IPFSClient.Dag().Pinning().AddMany(ctx, nodes); err != nil {
-		return fmt.Errorf("pinning set of nodes: %s", err)
-	}
 	var totalAddedSize int64
 	for _, n := range nodes {
-		s, err := n.Size()
+		s, err := n.Stat()
 		if err != nil {
 			return fmt.Errorf("getting size of node: %s", err)
 		}
-		totalAddedSize += int64(s)
+		totalAddedSize += int64(s.CumulativeSize)
+	}
+	currentBucketsSize, err := s.getBucketsTotalSize(ctx)
+	if err != nil {
+		return fmt.Errorf("getting current buckets total size: %s", err)
+	}
+
+	if s.BucketsTotalMaxSize > 0 && currentBucketsSize+totalAddedSize > s.BucketsTotalMaxSize {
+		return ErrBucketsTotalSizeExceedsMaxSize
+	}
+
+	if err := s.IPFSClient.Dag().Pinning().AddMany(ctx, nodes); err != nil {
+		return fmt.Errorf("pinning set of nodes: %s", err)
 	}
+
 	if err := s.sumBytesPinned(ctx, totalAddedSize); err != nil {
 		return fmt.Errorf("adding pinned size to account quota: %s", err)
 	}
 	return nil
 }
 
-// sumBytesPinned accounts the provided delta difference into the account/user
-// buckets total size quota.
+// sumBytesPinned adds the provided delta to the buckets total size from
+// the account/user.
 func (s *Service) sumBytesPinned(ctx context.Context, delta int64) error {
-	key, isUser := ownerFromContext(ctx)
-	if !isUser {
-		a, err := s.Collections.Accounts.Get(ctx, key)
-		if err != nil {
-			return fmt.Errorf("getting account from context: %s", err)
-		}
-		newTotalSize := a.BucketsTotalSize + delta
-		if err := s.Collections.Accounts.SetBucketsTotalSize(ctx, key, newTotalSize); err != nil {
+	a := accountFromContext(ctx)
+	if a != nil {
+		a.BucketsTotalSize = a.BucketsTotalSize + delta
```

Comment:
Update the account struct that lives in the `ctx` (not only send the update to Mongo)